### PR TITLE
Update AppBar and AppBar Theme to new Theme conventions and latest Material spec

### LINKED ISSUE #50606: AppBar doesn't apply overriden Theme
I have a very simple `AppBar` outside of a `Scaffold`, like:

```dart
class MyScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        children: <Widget>[
          AppBar(
            title: Text('hi'),
            flexibleSpace: Container(color: Colors.purple, height: 200,),
          ),
          Expanded(
            child: Center(
              child: Text('hi again'),
            ),
          ),
        ],
      ),
    );
  }
}
```

Which renders like:

image

The `MaterialApp` theme `primaryColorBrightness` is `Brightness.light`.

Now, if I change the `MaterialApp` theme `primaryColorBrightness` to `Brightness.dark` the status bar color AND the `AppBar` icons and text change to white:

image

But if I just want to change the `Theme` for  `MyScreen` and not the whole app, the status bar changes color to white, but the `AppBar` doesn't:

```dart
class MyScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Theme(
      data: Theme.of(context).copyWith(primaryColorBrightness: Brightness.dark),
      child: Scaffold(
        body: Column(
          children: <Widget>[
            AppBar(
              title: Text('hi'),
              flexibleSpace: Container(color: Colors.purple, height: 200,),
            ),
            Expanded(
              child: Center(
                child: Text('hi again'),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

As can be seen here:

image

This inconsistency seems like a bug to me.

### LINKED ISSUE #51820: Inconsistent AppBar color naming
The `AppBar` has a property called `backgroundColor`, but the `AppBarTheme` only has `color`, they appear to be the same as `backgroundColor` docs point out that it would be `AppBarTheme.color` if `backgroundColor` is null. 

I suggest that `AppBarTheme.color` be renamed to `AppBarTheme.backgroundColor` for the sake of consistency.

### LINKED ISSUE #61618: AppBar.brightness not changing the title widget text color
## Steps to Reproduce

1. Create this app:
```
import 'package:flutter/material.dart';

final Color darkBlue = Color.fromARGB(255, 18, 32, 47);

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(),
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        appBar: AppBar(
//           backgroundColor: Colors.black.withOpacity(.01),
//           brightness: Brightness.light,
        title: Text('App Name'),
          ),
        body: Center(
          child: MyWidget(),
        ),
      ),
    );
  }
}

class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Text('Hello, World!', style: Theme.of(context).textTheme.headline4);
  }
}

```
2. uncomment the two commented lines
3. The title text stays white

**Expected results:** 
 I expected the title text to go dark. The icons in the status bar of the phone go dark, but not the text of the title.

**Actual results:** 
The text in the title widget remains white
image

I understand that I can update the `primaryColorBrightness` in the main theme. However, it seems wrong to change the primary color brightness on the whole app just to adjust the color of an AppBar.

Am I over looking something simple?

### LINKED ISSUE #67497: AppBar ignores Brightness and AnnotatedRegion doesn't work
For what I could find in the docs `AppBar` should automatically set the status bar style according to it's `brightness`. Not only it fails to do so, but this behavior isn't documented on the `brightness` parameter, only on the `SystemChrome.setSystemUIOverlayStyle` (that I could find at least).

This code shows the issue:

```dart
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        appBarTheme: AppBarTheme(
          elevation: 0,
          color: Theme.of(context).canvasColor,
          brightness: Brightness.light,
        ),
      ),
      home: Scaffold(
        appBar: AppBar(title: Text('Test Text')),
        body: Center(
          child: Text('Content'),
        ),
      ),
    );
  }
}
```

And here's what we get: 

image

Same thing setting the `brightness` as `Brightness.dark`:

image

As you can see the `AppBar` `Text` is just white, no matter what the `brightness` is, also the Android Status Bar doesn't change colors and style at all. I even tried wrapping the `Scaffold` on a `Builder` as to see if it just couldn't find the `Theme` because of the `context`, but it didn't make any differences.

Adding a color scheme to the `ThemeData` does change the `Text` color, but this only shows one more time that the `AppBar`'s `brightness` isn't being used.

This is running on an Android AVD with Android API 29.

Now, further investigating this (as my main goal was to change the status bar color and icons color) I can't make `AnnotatedRegion` work:

```dart
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        appBarTheme: AppBarTheme(
          elevation: 0,
          color: Theme.of(context).canvasColor,
          brightness: Brightness.light,
        ),
      ),
      home: AnnotatedRegion(
        value: SystemUiOverlayStyle.dark.copyWith(
          statusBarColor: Colors.red,
          statusBarIconBrightness: Brightness.dark,
        ),
        child: Scaffold(
          appBar: AppBar(title: Text('Test Text')),
          body: Center(
            child: Text('Content'),
          ),
        ),
      ),
    );
  }
}
```

This is a standard `flutter create` App, nothing fancy on the Android side of things, I just changed the Dart code. I tried to set the `SystemUiOverlayStyle` in the `AnnotatedRegion` with `statusBarIconBrightness: Brightness.dark` and `statusBarIconBrightness: Brightness.light`, with no success.

Trying to use the "not recommended" `SystemChrome.setSystemUIOverlayStyle` works a bit:

```dart
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        appBarTheme: AppBarTheme(
          elevation: 0,
          color: Theme.of(context).canvasColor,
          brightness: Brightness.light,
        ),
      ),
      // home: MyHomePage(title: 'Flutter Demo Home Page'),
      home: Scaffold(
        appBar: AppBar(title: Text('Test Text')),
        body: Center(
          child: RaisedButton(
            child: Text('Content'),
            onPressed: () =>
                SystemChrome.setSystemUIOverlayStyle(SystemUiOverlayStyle.light),
          ),
        ),
      ),
    );
  }
}
```

Changing the `SystemChrome.setSystemUIOverlayStyle(SystemUiOverlayStyle.light)` to `SystemChrome.setSystemUIOverlayStyle(SystemUiOverlayStyle.dark)` makes no visual difference. But setting it to 

```dart
SystemChrome.setSystemUIOverlayStyle(SystemUiOverlayStyle.light.copyWith(statusBarColor: Colors.red))
```

Does change the Status Bar color to red, but I can't find any configuration that gets to change the Status Bar Icon color to a dark one. Bear in mind this code worked just fine on Flutter 1.20.4, but now it doesn't make any differences:

```dart
SystemChrome.setSystemUIOverlayStyle(
          SystemUiOverlayStyle.light.copyWith(
              statusBarColor: Colors.transparent,
              statusBarIconBrightness: Brightness.light),
        );
```

My experience with the `SystemUiOverlayStyle` throughout several Flutter versions have been pretty bad, sometimes it works and sometimes it doesn't, while we can't find much examples or documentations on this.

So to summarize:

* `AppBar` is ignoring its `brightness` parameter. 
* `AppBar` doesn't document how `brightness` affects the OS Status Bar (But `AnnotatedRegion` does make a reference to it.
* `AnnotatedRegion` doesn't change the Android Status Bar.
* `SystemChrome.setSystemUIOverlayStyle` partly changes the Android Status Bar.

<details>
  <summary>flutter doctor</summary>

```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 1.22.0, on Microsoft Windows [Version 10.0.19041.508], locale pt-BR)

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[√] Android Studio (version 4.0)
[√] IntelliJ IDEA Ultimate Edition (version 2019.3)
[√] VS Code (version 1.48.2)
[√] Connected device (1 available)

• No issues found!
```
</details>

### LINKED ISSUE #67921: [Material] AppBar background color should check for ColorScheme.primary
Currently, the AppBar checks for ThemeData.primary, but does not check to use ColorScheme.primary. When the backgroundColor is resolved, it should check to use ColorScheme.primary if its not null.

### LINKED ISSUE #70645: Remove support for configuring AppBar background color with ThemeData.primaryColor
[ThemeData.primaryColor](https://api.flutter.dev/flutter/material/ThemeData/primaryColor.html) is currently the default background color for AppBar. Apps should configure the default background color for AppBar's with AppBar.color, AppBarTheme.backgroundColor, or the theme's ColorScheme.primaryColor.

Support for ThemeData.primaryColor is being phased out because it's redundant with ColorScheme.primary and not as clear as AppBar.color or AppBarTheme.backgroundColor.

Support for ThemeData.primaryColor, when it's not the same as the ColorScheme's primary, has been temporarily enabled for AppBar in ... until old applications have had a chance to update.

### PR DESCRIPTION
**Pull request by HansMuller**
AppBar and AppBarTheme are changing to bring them into alignment with the Material spec and to make configuring AppBars relatively straightforward.

This pull request is the first step in a sequence of changes and is not intended to be a "breaking change" https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes. The temporary `backwardsCompatibility` AppBar flag can be set to false (default is true) to enable the changes.

## AppBar Changes

All of the properties listed here can be overridden with an AppBar widget parameter or by an AppBarTheme parameter with the same name. For example the default AppBar background color is the theme's ColorScheme.primary color https://api.flutter.dev/flutter/material/ColorScheme/primary.html. If AppBar.backgroundColor is non-null, then it's used instead of the default.  If AppBar.backgroundColor is null and AppBarTheme.backgroundColor is non-null then AppBarTheme.backgroundColor is used. By default both AppBar.backgroundColor and AppBarThemme.backgroundColor are null.

To keep the following descriptions brief and clear we use "or" to mean `??`, "theme" for the value of `Theme.of(context)`, "colorScheme" for the value of `Theme.of(context).colorScheme`, "appbarTheme" for the value of `AppBarTheme.of(context)`, and brightness for the value of `Theme.of(context).colorScheme.brightness`.

The app bar's "toolbar" is contains the app bar's leading, title, and actions widgets. The AppBar uses a NavigationToolbar to manage the toolbar's layout.

The following AppBar properties are either new, obsolete, or have changed.

### `backwardsCompatibility` (new temporary property)
Tests and apps can opt into the new properties and defaults by setting this property to false.

This property will exist until the Flutter repo and all of the existing tests have been migrated. After that it will be deprecated and eventually removed.

### `backgroundColor` (existing property, updated defaults)
The fill color to use for the app bar's [Material].

The default is appBarTheme.backgroundColor or colorScheme.surface if the theme's brightness is dark, colorScheme.primary if the theme's brightness is light.

If backwardsCompatible is true then the default is AppBarTheme.color or theme.primaryColor. By default, theme.primaryColor is darkColors.grey[900] if the theme's brightness dark, theme.colorScheme.primary otherwise.

The new default values bring this property into line with the Material spec. Using AppBarTheme.backgroundColor instead of AppBarTheme.color for the theme's version of the property is for the sake of consistency with all of the other AppBar properties and their AppBarTheme counterparts.  This default value's change also avoids the use of theme.primaryColor which confusingly is only the theme's primary color if the overall theme's brightness is light.

### `foregroundColor` (new property)
The color of [Text] and [Icon]s within the app bar's toolbar.

The default is appBarTheme.foregroundColor or colorScheme.onSurface if the theme's brightness is dark, colorScheme.onPrimary if the theme's brightness is light.

This property was added to simplify specifying the default text and icon color for the AppBar's toolbar. It is used to configure the default values of textStyle, titleTextStyle, iconTheme, and actionsIconTheme.

### `toolbarTextStyle` (new property)
The TextStyle for the leading and actions widgets.

The default is appBarTheme.toolbarTextStyle or theme.textTheme.bodyText2.copyWith(foregroundColor).

If backwardsCompatible is true then the default is appBarTheme.textTheme.bodyText2 or theme.primaryTextTheme.bodyText2.

This property was added to simplify setting the text style of the leading and actions widgets by eliminating the need to create a TextTheme. This change also avoids the use of the confusing theme.primaryTextTheme value whose color is neither the theme's primary ColorScheme color _nor_ the theme's primaryColor.

### `titleTextStyle` (new property)
The TextStyle for the title widget.

The default is appBarTheme.titleTextStyle or theme.textTheme?.headline6.copyWith(foregroundColor).

If backwardsCompatible is true then the default is appBarTheme.textTheme.headline6 or theme.primaryTextTheme.headline6.

This property was added to simplify setting the AppBar title's text style by eliminating the need to create a TextTheme. This change also avoids the use of the confusing theme.primaryTextTheme value whose color is neither the theme's primary ColorScheme color _nor_ the theme's primaryColor.

### `iconTheme` (existing property, updated default)
The color, opacity, and size for toolbar icons.

The default is theme.iconTheme.copyWith(color: foregroundColor).

If backwards compatible is true then the default value is theme.primaryIconTheme.

This property's default value was changed to simplify the common case of specifying the AppBar's overall text/icon foreground color. This change also avoids the use of the confusing theme.primaryIconTheme value whose color is neither the theme's primary ColorScheme color _nor_ the theme's primaryColor.

### `actionsIconTheme` (existing property, updated defaults)
The color, opacity, and size to use for the icons that appear in the app bar's actions.

Default is the iconTheme.

This property's effective default value was changed to simplify the common case of specifying the AppBar's overall text/icon foreground color. This change also avoids the use of the confusing theme.primaryIconTheme value whose color is neither the theme's primary ColorScheme color _nor_ the theme's primaryColor.

### `brightness` (obsolete property)
This property is no longer used.

For reasons lost to history the AppBarTheme's `brightness` parameter does not override the color scheme's brightness for the other AppBar properties. It's only ever been used to select one of two values for systemOverlayStyle. Specifying a systemOverlayStyle value is more easily and clearly specified by just providing an explicit value for the new overlayStyle property.

### `systemOverlayStyle` (new property)
The style to use for the system overlays that overlap the AppBar.

The default is appBarTheme.systemOverlayStyle or SystemUiOverlayStyle.light for dark themes, SystemUiOverlayStyle.dark otherwise.

This property was added because setting this value indirectly, via the (now obsolete) brightness property, was needlessly complicated and less flexible.

## AppBarTheme Changes

For more information, see the descriptions of the corresponding AppBar properties above.

### `backgroundColor` (new property)
The default value for the app bar's backgroundColor.

### `brightness` (obsolete property)
No longer used to select the app bar's systemOverlayStyle.

### `systemOverlayStyle` (new property)
The default value for the app bar's systemOverlayStyle.

## Fixes

Fixes https://github.com/flutter/flutter/issues/70645
Fixes https://github.com/flutter/flutter/issues/67921
Fixes https://github.com/flutter/flutter/issues/67497
Fixes https://github.com/flutter/flutter/issues/61618
Fixes https://github.com/flutter/flutter/issues/51820
Fixes https://github.com/flutter/flutter/issues/50606

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 4 untriaged digest(s) on patchset 7.
View them at https://flutter-gold.skia.org/cl/github/71184

**Comment by skia-gold** [comment]
Gold has detected about 4 untriaged digest(s) on patchset 8.
View them at https://flutter-gold.skia.org/cl/github/71184

**Comment by feinstein** [comment]
What's the difference/priority of `iconTheme` and `foregroundColor`?

**Comment by feinstein** [comment]
(just a small typo "ApBarTheme Changes" on the last topic)

**Comment by lig** [comment]
@HansMuller Thanks a lot! The behavior before this PR causes the only point of hacks singularity in my code.

**Comment by rydmike** [comment]
@HansMuller based on the summary above and quick look at code updates, it looks like it should solve multiple reported issues and challenges people have been having with the AppBar theming. **Nice work!** 👍🏻 

For my use case, a simple test will be if it is now easier (=possible) to create a working MaterialApp ThemeData object that uses a white app bar background color, when the light theme's primary color is of brightness dark. The white or light app bar background color then requires dark/black text and icon theme, which goes against typography color for the primaryTextTheme and the "on" primary color.

Or wise versa the case (but much less common and not really often used) in a dark app theme with a colored app bar with a background color of brightness light, that requires dark text/icon color, which goes against the color of the textTheme's typography color in dark themes.

Basically these two cases go against needed text and icon colors we get via the overall default light and dark themes as default for the app bar. Due to this conflict, those two cases were a bit complicated to setup via just the standard ThemeData factory before. **If this is now possible** with these new features, well then this is really golden, **based on the specs and the code it certainly looks like it should be.**

**Testing the above use cases** will just be a simple case of redoing the things I discussed here https://github.com/flutter/flutter/issues/50606#issuecomment-662216140, **using these new features** to verify that it now works without resorting to creating a new text theme, outside Flutter's text themes that does not get correct localizations applied.

I also appreciate that these new theming features brings the app bar theming in line with color scheme based themes and how they behave, which of course is according to the theme consistency roadmap.

I will test these new features as soon as they land in master. **Generally it looks good and is a very welcome improvement.** 
**Thanks Hans!** 💙

**Comment by HansMuller** [comment]
@feinstein 
> What's the difference/priority of iconTheme and foregroundColor?
If iconTheme is specified then its color is used, even if foregroundColor is also specified. If iconTheme is not specified, then foregroundColor is the default iconTheme's color. IconThemeData.color can be null so, if iconTheme specified and its color is null, the AppBar would inherit the enclosing IconTheme's color.  That's all the same as it ever was. I will make sure that there's a test which confirms as much.

**Comment by feinstein** [comment]
Perfect!

**Comment by HansMuller** [comment]
@rydmike 

I believe both of the use cases you brought up should be straightforward.

> For my use case, a simple test will be if it is now easier (=possible) to create a working MaterialApp ThemeData object that uses a white app bar background color, when the light theme's primary color is of brightness dark.

```dart
ThemeData.from(colorScheme: ColorScheme.dark()).copyWith(
  appBarTheme: AppBarTheme(
    backgroundColor: Colors.white,
    foregroundColor: Colors.black,
  ),
)
```

> a dark app theme with a colored app bar with a background color of brightness light, that requires dark text/icon color, which goes against the color of the textTheme's typography color in dark themes.

```dart
ThemeData.from(colorScheme: ColorScheme.dark()).copyWith(
  appBarTheme: AppBarTheme(
    backgroundColor: Colors.amber,
    foregroundColor: Colors.black,
  ),
)
```

**Comment by rydmike** [comment]
@HansMuller That's what  gathered from looking at the changes too. 
All tests appear to be green, can we look forward to a merge with master soon? Or is it still pending internal code review?

**Comment by HansMuller** [comment]
It still needs a review however it's substantially similar to an earlier PR, so hopefully the review will not take too long.

File: packages/flutter/lib/src/material/app_bar_theme.dart [hunk 1]
```diff
@@ -5,111 +5,191 @@
...
     this.textTheme,
     this.centerTitle,
     this.titleSpacing,
+    this.toolbarTextStyle,
+    this.titleTextStyle,
+    this.systemOverlayStyle,
   });
 
-  /// Default value for [AppBar.brightness].
+  /// This property is obsolete, please use [systemOverlayStyle] instead.
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/app_bar_theme.dart:47]: Should we mark this as deprecated, or are you waiting to get things migrated first?

File: packages/flutter/lib/src/material/app_bar_theme.dart [hunk 2]
```diff
@@ -5,111 +5,191 @@
...
   ///
-  /// If null, [AppBar] uses [ThemeData.primaryColorBrightness].
+  /// See also:
+  ///
+  ///  * [systemOverlayStyle], which overrides the default value of
+  ///    [AppBar.systemOverlayStyle] in all descendant [AppBar] widgets.
   final Brightness? brightness;
 
-  /// Default value for [AppBar.backgroundColor].
+  /// Obsolete property that overrides the default value of
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/app_bar_theme.dart:59]: Same question about deprecation here.

**Review by darrenaustin — APPROVED** [review]
Overall, looks awesome with a lot of cleaned up documentation and more consistent property usage. Just had a few minor questions and caught a typo, but otherwise LGTM!

File: packages/flutter/lib/src/material/app_bar.dart [hunk 3]
```diff
@@ -321,145 +346,324 @@ class AppBar extends StatefulWidget implements PreferredSizeWidget {
...
   /// See also:
   ///
   ///  * [PreferredSize], which can be used to give an arbitrary widget a preferred size.
   final PreferredSizeWidget? bottom;
 
+  /// {@template flutter.material.appbar.elevation}
   /// The z-coordinate at which to place this app bar relative to its parent.
   ///
-  /// This controls the size of the shadow below the app bar.
+  /// This proprty controls the size of the shadow below the app bar.
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/material/app_bar.dart:367]: Oops

File: packages/flutter/lib/src/material/app_bar.dart [hunk 4]
```diff
@@ -321,145 +346,324 @@ class AppBar extends StatefulWidget implements PreferredSizeWidget {
...
+  /// {@endtemplate}
   final double? leadingWidth;
 
+  /// {@template flutter.material.appbar.backwardsCompatibility}
+  /// If true, preserves the original defaults for the [backgroundColor],
+  /// [iconTheme], [actionsIconTheme] properties, and the original use of
+  /// the [textTheme] and [brightness] properties.
+  ///
+  /// This property is true by default.
+  /// {@endtemplate}
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/material/app_bar.dart:611]: Good point

File: packages/flutter/lib/src/material/app_bar.dart [hunk 5]
```diff
@@ -321,145 +346,324 @@ class AppBar extends StatefulWidget implements PreferredSizeWidget {
...
+
+  /// {@template flutter.material.appbar.systemOverlayStyle}
+  /// Specifies the style to use for the system overlays that overlap the AppBar.
+  ///
+  /// If this property is null, then [SystemUiOverlayStyle.light] is used if the
+  /// overall theme is dark, [SystemUiOverlayStyle.dark] otherwise. Theme brightness
+  /// is defined by [ColorScheme.brightness] for [ThemeData.colorScheme].
+  ///
+  /// The AppBar is built within a `AnnotatedRegion<SystemUiOverlayStyle>`
+  /// which causes [SystemChrome.setSystemUIOverlayStyle] to be called
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/material/app_bar.dart:657]: What I meant was that the widgets that the AppBar builds are created within an AnnotatedRegion. Will try and make that clearer.

File: packages/flutter/lib/src/material/app_bar.dart [hunk 6]
```diff
@@ -707,21 +939,19 @@ class _AppBarState extends State<AppBar> {
...
     }
-    final Brightness brightness = widget.brightness
-      ?? appBarTheme.brightness
-      ?? theme.primaryColorBrightness;
-    final SystemUiOverlayStyle overlayStyle = brightness == Brightness.dark
-      ? SystemUiOverlayStyle.light
-      : SystemUiOverlayStyle.dark;
+
+    final SystemUiOverlayStyle overlayStyle = widget.backwardsCompatibility
+      ? ((widget.brightness ?? appBarTheme.brightness ?? colorScheme.brightness) == Brightness.dark ? SystemUiOverlayStyle.light : SystemUiOverlayStyle.dark)
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/material/app_bar.dart:944]: OK, added a variable.

File: packages/flutter/lib/src/material/app_bar_theme.dart (hunk 2 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/app_bar_theme.dart:59]: I've tried to improve the documentation a little. for the obsolete properties, including this one.

**Pull request merged**